front/app/containers/Admin/reporting/containers/ReportBuilder/index.tsx

```diff
@@ -35,6 +36,46 @@ interface Props {
   reportLayout: ReportLayout;
 }
 
+const areEqual = (
+  craftjsCurrentData: SerializedNodes,
+  persistedData: SerializedNodes
+) => {
+  if (
+    isEmpty(persistedData) &&
+    (isEmpty(craftjsCurrentData) ||
+      craftjsCurrentData['ROOT'].nodes.length === 0)
+  ) {
+    return true;
+  }
+
+  // When the layout is saved, `undefined` fields are not saved
+  // (e.g., `parent` in { ROOT: { parent: undefined, ...}, ...} will be omitted).
+  // So, they are never present in reportLayout.attributes.craftjs_json,
+  // but they are present in the editor state `query.getSerializedNodes()`.
+  //
+  // The "for" loop does the same thing as this alternative version with JSON.stringify,
+  // but it's more performant.
+  //
+  // JSON.stringify emulates sending the data to the server and getting it back
+  // (it removes undefined values).
+  // JSON.parse makes sure that the comparison is not affected by the order of keys.
+  // return isEqual(
```

Comment:
Left this implementation here to make it easier to (1) understand how it works and (2) fix if `for key in` has some bugs.